spec/std/json/serialization_spec.cr

```diff
@@ -442,8 +575,74 @@ describe "JSON serialization" do
       {x: 1, y: "hello"}.to_json.should eq(%({"x":1,"y":"hello"}))
     end
 
-    it "does for Enum" do
-      JSONSpecEnum::One.to_json.should eq("1")
+    describe "Enum" do
+      it "normal enum" do
+        JSONSpecEnum::One.to_json.should eq %("one")
+        JSONSpecEnum.from_json(JSONSpecEnum::One.to_json).should eq(JSONSpecEnum::One)
+
+        JSONSpecEnum::OneHundred.to_json.should eq %("one_hundred")
+        JSONSpecEnum.from_json(JSONSpecEnum::OneHundred.to_json).should eq(JSONSpecEnum::OneHundred)
+
+        # undefined members can't be parsed back because the standard converter only accepts named
+        # members
+        JSONSpecEnum.new(42).to_json.should eq %("42")
+      end
+
+      it "flag enum" do
+        JSONSpecFlagEnum::One.to_json.should eq %(["one"])
+        JSONSpecFlagEnum.from_json(JSONSpecFlagEnum::One.to_json).should eq(JSONSpecFlagEnum::One)
+
+        JSONSpecFlagEnum::OneHundred.to_json.should eq %(["one_hundred"])
+        JSONSpecFlagEnum.from_json(JSONSpecFlagEnum::OneHundred.to_json).should eq(JSONSpecFlagEnum::OneHundred)
+
+        combined = JSONSpecFlagEnum::OneHundred | JSONSpecFlagEnum::One
+        combined.to_json.should eq %(["one","one_hundred"])
+        JSONSpecFlagEnum.from_json(combined.to_json).should eq(combined)
+
+        JSONSpecFlagEnum::None.to_json.should eq %([])
+        JSONSpecFlagEnum.from_json(JSONSpecFlagEnum::None.to_json).should eq(JSONSpecFlagEnum::None)
+
+        JSONSpecFlagEnum::All.to_json.should eq %(["one","two","one_hundred"])
+        JSONSpecFlagEnum.from_json(JSONSpecFlagEnum::All.to_json).should eq(JSONSpecFlagEnum::All)
+
+        JSONSpecFlagEnum.new(42).to_json.should eq %(["two"])
+      end
+    end
+
+    describe "Enum::NumberOrStringConverter" do
+      it "normal enum" do
+        converter = Enum::NumberOrStringConverter(JSONSpecEnum)
```

Comment:
To me it's a bit strange that it's number or string for reading, but just number for writing. Maybe we should just let it work with numbers? I can't imagine why one would like to use sometimes strings and sometimes numbers for reading, but just numbers for writing. It should always be either always numbers or always strings.

If it's because of backwards compatibility, maybe that could be in a shard.

Reply:
Yes, the idea was to have a simple converter for an easy migration from the current behaviour. You can just plug in `Enum::NumberOrStringConverter` and it works.

I agree that a more strict number converter would probably be better. But we didn't have that until now, so I would see that as an additional feature.

If we accept a reduced backward compatibility, we could leave `NumberOrStringConverter` out and add a strict `NumberConverter` directly for the main use case.

Reply:
I like the path to be more strict. I though that number or string was introduced as an option in the former PR.

Let's introduce `Enum::NumberConverter` and leave the default as proposed to string / List of string.

Reply:
Deserialization from number or string has been there since `Enum.from_json` was implemented. I wouldn't expect the string variant to be used much, because serialization was only as number.

So I'll change the converter to a strict `NumberConverter`.

If there's some actual demand, we could consider publishing `NumberOrStringConverter` as a shard or whatever, but it's really trivial to implement. And as I said, I don't expect much use anyways.

Reply:
I've changed the name to more generalized `ValueConverter`. I think that's better for describing the purpose.